src/L1/rollup/ScrollChain.sol

```diff
@@ -713,18 +907,22 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
         uint256 _length;
         if (version == 0) {
             (batchPtr, _length) = BatchHeaderV0Codec.loadAndValidate(_batchHeader);
-            _batchHash = BatchHeaderV0Codec.computeBatchHash(batchPtr, _length);
-            _batchIndex = BatchHeaderV0Codec.getBatchIndex(batchPtr);
-        } else if (version >= 1) {
+        } else if (version <= 2) {
             (batchPtr, _length) = BatchHeaderV1Codec.loadAndValidate(_batchHeader);
-            _batchHash = BatchHeaderV1Codec.computeBatchHash(batchPtr, _length);
-            _batchIndex = BatchHeaderV1Codec.getBatchIndex(batchPtr);
+        } else if (version >= 3) {
+            (batchPtr, _length) = BatchHeaderV3Codec.loadAndValidate(_batchHeader);
         }
+
+        // the code for compute batch hash is the same for V0, V1, V2, V3
+        // also the `_batchIndex` and `_totalL1MessagesPoppedOverall`.
+        _batchHash = BatchHeaderV0Codec.computeBatchHash(batchPtr, _length);
+        _batchIndex = BatchHeaderV0Codec.getBatchIndex(batchPtr);
+        _totalL1MessagesPoppedOverall = BatchHeaderV0Codec.getTotalL1MessagePopped(batchPtr);
```

Comment:
Very small nitpick, putting this below `revert ErrorIncorrectBatchHash();` could save some gas usage.